proposals/vocabularies.md

```diff
@@ -0,0 +1,267 @@
+specification.
+
+A vocabulary is identified by an absolute URI and is used to define a set of
+keywords. A vocabulary is generally defined in a human-readable _vocabulary
+description document_. (The URI for the vocabulary may be the same as the URL of
+where this vocabulary description document can be found, but no recommendation
+is made either for or against this practice.)
+
+A new keyword, `$vocabulary`, will be introduced into the Core specification as
+well. This keyword's value is an object with vocabulary URIs as keys and
+booleans as values. This keyword only has meaning within a meta-schema. A
+meta-schema which includes a vocabulary's URI in its `$vocabulary` keyword is
+said to "include" that vocabulary.
+
+```jsonc
+{
+  "$schema": "https://example.org/draft/next/schema",
+  "$id": "https://example.org/schema",
+  "$vocabulary": {
+    "https://example.org/vocab/vocab1": true,
+    "https://example.org/vocab/vocab2": true,
+    "https://example.org/vocab/vocab3": false
+  },
+  // ...
+}
+```
+
+A dialect is the set of vocabularies listed by a meta-schema. It is ephemeral
+and carries no identifier.
+
+_**NOTE** It is possible for two meta-schemas, which would have different `$id`
+values, to share a common dialect if they both declare the same set of
+vocabularies._
+
+A schema that declares a meta-schema (via `$schema`) which contains
+`$vocabulary` is declaring that only those keywords defined by the included
+vocabularies are to be processed when evaluating the schema. All other keywords
+are to be considered "unknown" and handled accordingly.
+
+The boolean values in `$vocabulary` signify implementation requirements for each
+vocabulary.
+
+- A `true` value indicates that the implementation must recognize the vocabulary
+  and be able to process each of the keywords defined it. If an implementation
+  does not recognize the vocabulary or cannot process all of its defined
+  keywords, the implementation must refuse to process the schema. These
+  vocabularies are also known as "required" vocabularies.
+- A `false` value indicates that the implementation is not required to recognize
+  the vocabulary or its keywords and may continue processing the schema anyway.
+  However, keywords that are not recognized or supported must be considered
+  "unknown" and handled accordingly. These vocabularies are also known as
+  "optional" vocabularies.
+
+Typically, but not required, a schema will accompany the vocabulary description
+document. This _vocabulary schema_ should carry an `$id` value which is distinct
+from the vocabulary URI. The purpose of the vocabulary schema is to provide
+syntactic validation for the the vocabulary's keywords' values for when the
+schema is being validated by a meta-schema that includes the vocabulary. (A
+vocabulary schema is not itself a meta-schema since it does not validate entire
+schemas.) To facilitate this extra validation, when a vocabulary schema is
```

Comment:
He didn't update the spec or anything.  This actually came out of reviewing @jviotti's book and trying to rework my vocab schemas.  I wrote about it [here](https://blog.json-everything.net/posts/updating-vocabs/), and we pulled out `$vocabulary` from them in #1460, #1461, and #1462.

They're not meta-schemas because they don't themselves describe full schemas; they are used by meta-schemas as components.

Reply:
I don't think we're quite talking about the same thing. I entirely agree that vocabulary meta-schemas shouldn't include the `$vocabulary` keyword.

> They're not meta-schemas because they don't themselves describe full schemas they are used by meta-schemas as components.

This distinction is what doesn't sit right with me. The way I see it, a component of a meta-schema is still a meta-schema. [Section 8.1.2.2](https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-00#rfc.section.8.1.2.2) is what I thought you were referring to here. Is that correct?

It does seem like the wording there is not considering schemas referenced by a meta-schema a meta-schema, but that's never how I've understood the word or how we've defined the word. I've always used the terms dialect meta-schema and vocabulary meta-schema. The `$vocabulary` keyword is only meaningful when the meta-schema is used as a dialect meta-schema.

Reply:
No, I wasn't referring to that section.

I'm looking at [Core 4.3.4](https://json-schema.org/draft/2020-12/json-schema-core#section-4.3.4) which _defines_ "meta-schema":

> A schema that itself describes a schema is called a meta-schema.

A vocabulary schema doesn't describe a schema, therefore it's not a meta-schema.

You wouldn't use the Meta-data vocab schema as a meta-schema; you reference it from a meta-schema.

Reply:
Interesting. We're using the same definition, but interpreting it differently. The way I see it, a vocabulary schema is validating a schema. It validates the syntax of keywords in a schema. I don't think the definition implies that a schema is only a meta-schema if it describes the _entire_ dialect used by the schema.

Reply:
It _is_ validating keyword usage inside of a schema, but it is not validating a schema.  It's a small distinction, but I think it's worth noting.

Calling the vocabulary schema a "meta-schema" is what led me to create my first vocab schemas _as_ meta-schemas, which multiple people recognized as the wrong way to do it.  I didn't understand why until Juan started to dig into it for his book.  I realized my mistake and started calling it a "vocab schema" from then on.  If _I_ got confused by this, others doubtlessly will (or have already) as well.  I just want the distinction to be apparent that this isn't used the in the same way that a full meta-schema would be.

A meta-schema is something that you'd reference with `$schema`.  You wouldn't put a vocabulary schema ID in `$schema`; its purpose is different.

Reply:
I had the same realisation while implementing vocab support - which leads to another question. Do we need to publish a canonical vocab metaschema schema?

Reply:
> Do we need to publish a canonical vocab metaschema schema?

I'm not sure what you're leading toward.  Are you questioning the current practice of defining a vocab schema?  Or are you saying that we should have a vocab schema that _can_ function as a meta-schema?

Reply:
> If _I_ got confused by this, others doubtlessly will (or have already) as well.

I 100% agree that we need to explain this better and using more descriptive terms is a big part of doing that well. But, the way you're currently expressing this is confusing to me because this doesn't mesh with my (and I'm sure others) understanding of the term "meta-schema". That's why I'd prefer to address this by introducing new terms that are more specific. Earlier I mention that I use the terms "dialect meta-schema" and "vocabulary meta-schema", but since the term "meta-schema" appears to be inconsistently understood, perhaps "dialect schema" and "vocabulary schema" is better.

Reply:
> I would not have made the mistake if we'd had a "vocab metaschema" schema that clearly defined what was allowed in a vocab metaschema. It would have been apparent that we were talking about a very similar, but different entity.

What would be the difference between these two meta-schemas? Everything that's allowed in a vocab schema is allowed in a dialect schema. Everything that's allowed in a dialect schema is allowed in a vocab schema. The only difference I can see would be their identifier.

Reply:
I agree with @jdesrosiers on this:  syntactically they're just schemas.

The only difference is intent.  A meta-schema is intended to describe/validate a schema.  It can _use_ vocab schemas (by reference) to accomplish that task.  A vocab schema on its own does not describe a schema; it's merely a building block in a meta-schema.  This is the distinction I'm making.  (One _could_ create a single-vocab meta-schema.)

I don't think having a separate meta-schema for vocab schemas is needed since anything you do in a general schema should technically be allowed in a vocab schema.  I'm not sure how _useful_ all of those features would be, but I don't see a reason to forbid any features.